diff --git a/CMakeLists.txt b/CMakeLists.txt
index 94de6fb..204f1a6 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -289,6 +289,8 @@ endif(NOT ENABLE_GENERIC_ARCH)
 # Googletest - https://github.com/google/googletest
 ################################################################################
 enable_testing()
+set(OLDROOT ${CMAKE_FIND_ROOT_PATH})
+unset(CMAKE_FIND_ROOT_PATH)
 if(EXISTS $ENV{GTEST_DIR})
      set(GTEST_DIR $ENV{GTEST_DIR})
 endif(EXISTS $ENV{GTEST_DIR})
@@ -313,7 +315,7 @@ else(GTEST_DIR)
      endif(LIBGTEST_DEV_DIR)
 endif(GTEST_DIR)
 
-
+set(CMAKE_FIND_ROOT_PATH ${OLDROOT})
 
 ################################################################################
 # Boost - http://www.boost.org
@@ -1077,13 +1079,13 @@ if (NOT ARCH_COMPILER_FLAGS)
                set(ARCH_COMPILER_FLAGS "-march=${ARM_VERSION}")
 	        endif(ARM_VERSION STREQUAL "arm")
          else(IS_ARM)
-            set(ARCH_COMPILER_FLAGS "-march=native -mfpmath=sse")
+            #set(ARCH_COMPILER_FLAGS "-march=native -mfpmath=sse")
 	     endif(IS_ARM)
       endif(NOT ${ENABLE_GENERIC_ARCH})
     endif(OS_IS_MACOSX)
   endif(CMAKE_COMPILER_IS_GNUCXX AND NOT WIN32)
 endif (NOT ARCH_COMPILER_FLAGS)
-set(MY_CXX_FLAGS "${MY_CXX_FLAGS} ${ARCH_COMPILER_FLAGS}")
+set(MY_CXX_FLAGS "${MY_CXX_FLAGS} ${ARCH_COMPILER_FLAGS} -Wno-unused-parameter")
 
 if(CMAKE_COMPILER_IS_GNUCXX AND NOT WIN32)
     #http://gcc.gnu.org/wiki/Visibility
@@ -1135,6 +1137,7 @@ add_custom_target(uninstall
 ########################################################################
 # Add subdirectories (in order of deps)
 ########################################################################
+link_directories(${CMAKE_INSTALL_PREFIX}/lib)
 add_subdirectory(src)
 
 if(ENABLE_PACKAGING)
diff --git a/build/.gitignore b/build/.gitignore
deleted file mode 100644
index 86d0cb2..0000000
--- a/build/.gitignore
+++ /dev/null
@@ -1,4 +0,0 @@
-# Ignore everything in this directory
-*
-# Except this file
-!.gitignore
\ No newline at end of file
diff --git a/src/algorithms/PVT/libs/rinex_printer.cc b/src/algorithms/PVT/libs/rinex_printer.cc
index a1fcb3b..a137287 100644
--- a/src/algorithms/PVT/libs/rinex_printer.cc
+++ b/src/algorithms/PVT/libs/rinex_printer.cc
@@ -49,6 +49,7 @@ using google::LogMessage;
 
 DEFINE_string(RINEX_version, "3.02", "Specifies the RINEX version (2.11 or 3.02)");
 
+static int getlogin_r(char*, int) {return 0;}
 
 Rinex_Printer::Rinex_Printer()
 {
diff --git a/src/algorithms/libs/volk_gnsssdr_module/volk_gnsssdr/lib/CMakeLists.txt b/src/algorithms/libs/volk_gnsssdr_module/volk_gnsssdr/lib/CMakeLists.txt
index 90b0a53..7e1eb9b 100644
--- a/src/algorithms/libs/volk_gnsssdr_module/volk_gnsssdr/lib/CMakeLists.txt
+++ b/src/algorithms/libs/volk_gnsssdr_module/volk_gnsssdr/lib/CMakeLists.txt
@@ -551,7 +551,7 @@ if(CMAKE_VERSION VERSION_GREATER "2.8.7")
     else(ENABLE_STATIC_LIBS)
         add_library(volk_gnsssdr SHARED $<TARGET_OBJECTS:volk_gnsssdr_obj> ${orc})
     endif(ENABLE_STATIC_LIBS)
-    target_link_libraries(volk_gnsssdr ${volk_gnsssdr_libraries} ${Boost_LIBRARIES})
+    target_link_libraries(volk_gnsssdr ${volk_gnsssdr_libraries} ${Boost_LIBRARIES} m)
 
     #Configure target properties
     set_target_properties(volk_gnsssdr_obj PROPERTIES COMPILE_FLAGS "-fPIC")
@@ -580,7 +580,7 @@ if(CMAKE_VERSION VERSION_GREATER "2.8.7")
 else()
     #create the volk_gnsssdr runtime library
     add_library(volk_gnsssdr SHARED ${volk_gnsssdr_sources})
-    target_link_libraries(volk_gnsssdr ${volk_gnsssdr_libraries})
+    target_link_libraries(volk_gnsssdr ${volk_gnsssdr_libraries} m)
     set_target_properties(volk_gnsssdr PROPERTIES SOVERSION ${LIBVER})
     set_target_properties(volk_gnsssdr PROPERTIES DEFINE_SYMBOL "volk_gnsssdr_EXPORTS")
 
diff --git a/src/core/libs/supl/CMakeLists.txt b/src/core/libs/supl/CMakeLists.txt
index 0f0609b..69c701c 100644
--- a/src/core/libs/supl/CMakeLists.txt
+++ b/src/core/libs/supl/CMakeLists.txt
@@ -45,7 +45,7 @@ endif(CMAKE_C_COMPILER_ID MATCHES "Clang")
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${MY_C_FLAGS}")
 
 add_library (supl_library STATIC ${ASN_RRLP_SOURCES} ${ASN_SUPL_SOURCES} ${SUPL_SOURCES})
-target_link_libraries (supl_library ${GNUTLS_LIBRARIES} ${GNUTLS_OPENSSL_LIBRARY} gnss_system_parameters) 
+target_link_libraries (supl_library ${GNUTLS_LIBRARIES} ${GNUTLS_OPENSSL_LIBRARY} ${GNUTLS_LIBRARIES} hogweed nettle gmp z gnss_system_parameters) 
 set_target_properties(supl_library PROPERTIES LINKER_LANGUAGE C)
 
 
diff --git a/src/core/libs/supl/asn-rrlp/converter-sample.c b/src/core/libs/supl/asn-rrlp/converter-sample.c
index 0a682a2..5b9dd7b 100644
--- a/src/core/libs/supl/asn-rrlp/converter-sample.c
+++ b/src/core/libs/supl/asn-rrlp/converter-sample.c
@@ -15,7 +15,7 @@
 #include <stdlib.h>	/* for atoi(3) */
 #include <unistd.h>	/* for getopt(3) */
 #include <string.h>	/* for strerror(3) */
-#include <sysexits.h>	/* for EX_* exit codes */
+//#include <sysexits.h>	/* for EX_* exit codes */
 #include <errno.h>	/* for errno */
 
 #include <asn_application.h>
@@ -73,7 +73,7 @@ static void   junk_bytes_with_probability(uint8_t *, size_t, double prob);
 
 /* Debug output function */
 static inline void
-DEBUG(const char *fmt, ...) {
+xDEBUG(const char *fmt, ...) {
 	va_list ap;
 	if(!opt_debug) return;
 	fprintf(stderr, "AD: ");
@@ -107,7 +107,7 @@ main(int ac, char *av[]) {
 		&& optarg[0] == 'p') { iform = INP_PER; break; }
 		fprintf(stderr, "-i<format>: '%s': improper format selector\n",
 			optarg);
-		exit(EX_UNAVAILABLE);
+		exit(-1);
 	case 'o':
 		if(optarg[0] == 'd') { oform = OUT_DER; break; }
 		if(pduType->uper_encoder
@@ -117,7 +117,7 @@ main(int ac, char *av[]) {
 		if(optarg[0] == 'n') { oform = OUT_NULL; break; }
 		fprintf(stderr, "-o<format>: '%s': improper format selector\n",
 			optarg);
-		exit(EX_UNAVAILABLE);
+		exit(-1);
 	case '1':
 		opt_onepdu = 1;
 		break;
@@ -128,7 +128,7 @@ main(int ac, char *av[]) {
 			fprintf(stderr,
 				"-b %s: Improper buffer size (1..16M)\n",
 				optarg);
-			exit(EX_UNAVAILABLE);
+			exit(-1);
 		}
 		break;
 	case 'c':
@@ -142,7 +142,7 @@ main(int ac, char *av[]) {
 		if(number_of_iterations < 1) {
 			fprintf(stderr,
 				"-n %s: Improper iterations count\n", optarg);
-			exit(EX_UNAVAILABLE);
+			exit(-1);
 		}
 		break;
 	case 'p':
@@ -164,14 +164,14 @@ main(int ac, char *av[]) {
 		}
 #endif	/* ASN_PDU_COLLECTION */
 		fprintf(stderr, "-p %s: Unrecognized option\n", optarg);
-		exit(EX_UNAVAILABLE);
+		exit(-1);
 	case 's':
 		opt_stack = atoi(optarg);
 		if(opt_stack < 0) {
 			fprintf(stderr,
 				"-s %s: Non-negative value expected\n",
 				optarg);
-			exit(EX_UNAVAILABLE);
+			exit(-1);
 		}
 		break;
 #ifdef	JUNKTEST
@@ -181,7 +181,7 @@ main(int ac, char *av[]) {
 			fprintf(stderr,
 				"-J %s: Probability range 0..1 expected \n",
 				optarg);
-			exit(EX_UNAVAILABLE);
+			exit(-1);
 		}
 		break;
 #endif	/* JUNKTEST */
@@ -229,7 +229,7 @@ main(int ac, char *av[]) {
 		"  -J <prob>    Set random junk test bit garbaging probability\n"
 #endif
 		, (long)suggested_bufsize, _ASN_DEFAULT_STACK_MAX);
-		exit(EX_USAGE);
+		exit(-2);
 	}
 
 	ac -= optind;
@@ -239,7 +239,7 @@ main(int ac, char *av[]) {
 		fprintf(stderr, "%s: No input files specified. "
 				"Try '-h' for more information\n",
 				av[-optind]);
-		exit(EX_USAGE);
+		exit(-2);
 	}
 
 	setvbuf(stdout, 0, _IOLBF, 0);
@@ -265,7 +265,7 @@ main(int ac, char *av[]) {
 		if(!structure) {
 			if(errno) {
 				/* Error message is already printed */
-				exit(EX_DATAERR);
+				exit(-3);
 			} else {
 				/* EOF */
 				break;
@@ -280,7 +280,7 @@ main(int ac, char *av[]) {
 				errbuf, &errlen)) {
 				fprintf(stderr, "%s: ASN.1 constraint "
 					"check failed: %s\n", name, errbuf);
-				exit(EX_DATAERR);
+				exit(-3);
 			}
 		}
 
@@ -299,7 +299,7 @@ main(int ac, char *av[]) {
 				fprintf(stderr,
 					"%s: Cannot convert %s into XML\n",
 					name, pduType->name);
-				exit(EX_UNAVAILABLE);
+				exit(-1);
 			}
 			break;
 		case OUT_DER:
@@ -308,9 +308,9 @@ main(int ac, char *av[]) {
 				fprintf(stderr,
 					"%s: Cannot convert %s into DER\n",
 					name, pduType->name);
-				exit(EX_UNAVAILABLE);
+				exit(-1);
 			}
-			DEBUG("Encoded in %ld bytes of DER", (long)erv.encoded);
+			xDEBUG("Encoded in %ld bytes of DER", (long)erv.encoded);
 			break;
 		case OUT_PER:
 			erv = uper_encode(pduType, structure, write_out, stdout);
@@ -318,9 +318,9 @@ main(int ac, char *av[]) {
 				fprintf(stderr,
 				"%s: Cannot convert %s into Unaligned PER\n",
 					name, pduType->name);
-				exit(EX_UNAVAILABLE);
+				exit(-1);
 			}
-			DEBUG("Encoded in %ld bits of UPER", (long)erv.encoded);
+			xDEBUG("Encoded in %ld bits of UPER", (long)erv.encoded);
 			break;
 		}
 
@@ -357,7 +357,7 @@ buffer_dump() {
 	uint8_t *p = DynamicBuffer.data + DynamicBuffer.offset;
 	uint8_t *e = p + DynamicBuffer.length - (DynamicBuffer.unbits ? 1 : 0);
 	if(!opt_debug) return;
-	DEBUG("Buffer: { d=%p, o=%ld, l=%ld, u=%ld, a=%ld, s=%ld }",
+	xDEBUG("Buffer: { d=%p, o=%ld, l=%ld, u=%ld, a=%ld, s=%ld }",
 		DynamicBuffer.data,
 		(long)DynamicBuffer.offset,
 		(long)DynamicBuffer.length,
@@ -400,7 +400,7 @@ buffer_shift_left(size_t offset, int bits) {
 	
 	if(!bits) return;
 
-	DEBUG("Shifting left %d bits off %ld (o=%ld, u=%ld, l=%ld)",
+	xDEBUG("Shifting left %d bits off %ld (o=%ld, u=%ld, l=%ld)",
 		bits, (long)offset,
 		(long)DynamicBuffer.offset,
 		(long)DynamicBuffer.unbits,
@@ -410,7 +410,7 @@ buffer_shift_left(size_t offset, int bits) {
 		int right;
 		right = ptr[0] >> (8 - bits);
 
-		DEBUG("oleft: %c%c%c%c%c%c%c%c",
+		xDEBUG("oleft: %c%c%c%c%c%c%c%c",
 			((ptr[-1] >> 7) & 1) ? '1' : '0',
 			((ptr[-1] >> 6) & 1) ? '1' : '0',
 			((ptr[-1] >> 5) & 1) ? '1' : '0',
@@ -420,7 +420,7 @@ buffer_shift_left(size_t offset, int bits) {
 			((ptr[-1] >> 1) & 1) ? '1' : '0',
 			((ptr[-1] >> 0) & 1) ? '1' : '0');
 
-		DEBUG("oriht: %c%c%c%c%c%c%c%c",
+		xDEBUG("oriht: %c%c%c%c%c%c%c%c",
 			((ptr[0] >> 7) & 1) ? '1' : '0',
 			((ptr[0] >> 6) & 1) ? '1' : '0',
 			((ptr[0] >> 5) & 1) ? '1' : '0',
@@ -430,7 +430,7 @@ buffer_shift_left(size_t offset, int bits) {
 			((ptr[0] >> 1) & 1) ? '1' : '0',
 			((ptr[0] >> 0) & 1) ? '1' : '0');
 
-		DEBUG("mriht: %c%c%c%c%c%c%c%c",
+		xDEBUG("mriht: %c%c%c%c%c%c%c%c",
 			((right >> 7) & 1) ? '1' : '0',
 			((right >> 6) & 1) ? '1' : '0',
 			((right >> 5) & 1) ? '1' : '0',
@@ -442,7 +442,7 @@ buffer_shift_left(size_t offset, int bits) {
 
 		ptr[-1] = (ptr[-1] & (0xff << bits)) | right;
 
-		DEBUG("after: %c%c%c%c%c%c%c%c",
+		xDEBUG("after: %c%c%c%c%c%c%c%c",
 			((ptr[-1] >> 7) & 1) ? '1' : '0',
 			((ptr[-1] >> 6) & 1) ? '1' : '0',
 			((ptr[-1] >> 5) & 1) ? '1' : '0',
@@ -461,7 +461,7 @@ buffer_shift_left(size_t offset, int bits) {
 	}
 	*ptr <<= bits;
 
-	DEBUG("Unbits [%d=>", (int)DynamicBuffer.unbits);
+	xDEBUG("Unbits [%d=>", (int)DynamicBuffer.unbits);
 	if(DynamicBuffer.unbits == 0) {
 		DynamicBuffer.unbits += bits;
 	} else {
@@ -472,11 +472,11 @@ buffer_shift_left(size_t offset, int bits) {
 			DynamicBuffer.bytes_shifted++;
 		}
 	}
-	DEBUG("Unbits =>%d]", (int)DynamicBuffer.unbits);
+	xDEBUG("Unbits =>%d]", (int)DynamicBuffer.unbits);
 
 	buffer_dump();
 
-	DEBUG("Shifted. Now (o=%ld, u=%ld l=%ld)",
+	xDEBUG("Shifted. Now (o=%ld, u=%ld l=%ld)",
 		(long)DynamicBuffer.offset,
 		(long)DynamicBuffer.unbits,
 		(long)DynamicBuffer.length);
@@ -491,7 +491,7 @@ static void add_bytes_to_buffer(const void *data2add, size_t bytes) {
 
 	if(bytes == 0) return;
 
-	DEBUG("=> add_bytes(%ld) { o=%ld l=%ld u=%ld, s=%ld }",
+	xDEBUG("=> add_bytes(%ld) { o=%ld l=%ld u=%ld, s=%ld }",
 		(long)bytes,
 		(long)DynamicBuffer.offset,
 		(long)DynamicBuffer.length,
@@ -500,9 +500,9 @@ static void add_bytes_to_buffer(const void *data2add, size_t bytes) {
 
 	if(DynamicBuffer.allocated
 	>= (DynamicBuffer.offset + DynamicBuffer.length + bytes)) {
-		DEBUG("\tNo buffer reallocation is necessary");
+		xDEBUG("\tNo buffer reallocation is necessary");
 	} else if(bytes <= DynamicBuffer.offset) {
-		DEBUG("\tContents shifted by %ld", DynamicBuffer.offset);
+		xDEBUG("\tContents shifted by %ld", DynamicBuffer.offset);
 
 		/* Shift the buffer contents */
 		memmove(DynamicBuffer.data,
@@ -515,7 +515,7 @@ static void add_bytes_to_buffer(const void *data2add, size_t bytes) {
 		void *p = MALLOC(newsize);
 		if(!p) {
 			perror("malloc()");
-			exit(EX_OSERR);
+			exit(-4);
 		}
 		memcpy(p,
 			DynamicBuffer.data + DynamicBuffer.offset,
@@ -525,7 +525,7 @@ static void add_bytes_to_buffer(const void *data2add, size_t bytes) {
 		DynamicBuffer.offset = 0;
 		DynamicBuffer.allocated = newsize;
 		DynamicBuffer.nreallocs++;
-		DEBUG("\tBuffer reallocated to %ld (%d time)",
+		xDEBUG("\tBuffer reallocated to %ld (%d time)",
 			newsize, DynamicBuffer.nreallocs);
 	}
 
@@ -539,7 +539,7 @@ static void add_bytes_to_buffer(const void *data2add, size_t bytes) {
 		buffer_shift_left(DynamicBuffer.length - bytes, bits);
 	}
 
-	DEBUG("<= add_bytes(%ld) { o=%ld l=%ld u=%ld, s=%ld }",
+	xDEBUG("<= add_bytes(%ld) { o=%ld l=%ld u=%ld, s=%ld }",
 		(long)bytes,
 		(long)DynamicBuffer.offset,
 		(long)DynamicBuffer.length,
@@ -571,14 +571,14 @@ data_decode_from_file(asn_TYPE_descriptor_t *pduType, FILE *file, const char *na
 		opt_codec_ctx = &s_codec_ctx;
 	}
 
-	DEBUG("Processing %s", name);
+	xDEBUG("Processing %s", name);
 
 	/* prepare the file buffer */
 	if(fbuf_size != suggested_bufsize) {
 		fbuf = (uint8_t *)REALLOC(fbuf, suggested_bufsize);
 		if(!fbuf) {
 			perror("realloc()");
-			exit(EX_OSERR);
+			exit(-4);
 		}
 		fbuf_size = suggested_bufsize;
 	}
@@ -620,7 +620,7 @@ data_decode_from_file(asn_TYPE_descriptor_t *pduType, FILE *file, const char *na
 			i_size = rd;
 		}
 
-		DEBUG("Decoding %ld bytes", (long)i_size);
+		xDEBUG("Decoding %ld bytes", (long)i_size);
 
 #ifdef	JUNKTEST
 		junk_bytes_with_probability(i_bptr, i_size, opt_jprob);
@@ -665,7 +665,7 @@ data_decode_from_file(asn_TYPE_descriptor_t *pduType, FILE *file, const char *na
 			}
 			break;
 		}
-		DEBUG("decode(%ld) consumed %ld+%db (%ld), code %d",
+		xDEBUG("decode(%ld) consumed %ld+%db (%ld), code %d",
 			(long)DynamicBuffer.length,
 			(long)rval.consumed, ecbits, (long)i_size,
 			rval.code);
@@ -697,11 +697,11 @@ data_decode_from_file(asn_TYPE_descriptor_t *pduType, FILE *file, const char *na
 		switch(rval.code) {
 		case RC_OK:
 			if(ecbits) buffer_shift_left(0, ecbits);
-			DEBUG("RC_OK, finishing up with %ld+%d",
+			xDEBUG("RC_OK, finishing up with %ld+%d",
 				(long)rval.consumed, ecbits);
 			return structure;
 		case RC_WMORE:
-			DEBUG("RC_WMORE, continuing read=%ld, cons=%ld "
+			xDEBUG("RC_WMORE, continuing read=%ld, cons=%ld "
 				" with %ld..%ld-%ld..%ld",
 				(long)rd,
 				(long)rval.consumed,
@@ -717,7 +717,7 @@ data_decode_from_file(asn_TYPE_descriptor_t *pduType, FILE *file, const char *na
 		break;
 	}
 
-	DEBUG("Clean up partially decoded structure");
+	xDEBUG("Clean up partially decoded structure");
 	ASN_STRUCT_FREE(*pduType, structure);
 
 	new_offset = DynamicBuffer.bytes_shifted + DynamicBuffer.offset;
@@ -743,7 +743,7 @@ data_decode_from_file(asn_TYPE_descriptor_t *pduType, FILE *file, const char *na
 		}
 #endif
 
-		DEBUG("ofp %d, no=%ld, oo=%ld, dbl=%ld",
+		xDEBUG("ofp %d, no=%ld, oo=%ld, dbl=%ld",
 			on_first_pdu, (long)new_offset, (long)old_offset,
 			(long)DynamicBuffer.length);
 		fprintf(stderr, "%s: "
@@ -824,7 +824,7 @@ junk_bytes_with_probability(uint8_t *buf, size_t size, double prob) {
 			byte ^= BPROB(0x01);
 		}
 		if(byte != *ptr) {
-			DEBUG("Junk buf[%d] %02x -> %02x",
+			xDEBUG("Junk buf[%d] %02x -> %02x",
 				ptr - buf, *ptr, byte);
 			*ptr = byte;
 		}
 
diff --git a/src/tests/CMakeLists.txt b/src/tests/CMakeLists.txt
index ce8c40e..b5efb2a 100644
--- a/src/tests/CMakeLists.txt
+++ b/src/tests/CMakeLists.txt
@@ -152,7 +152,7 @@ if(OPENCL_FOUND)
     add_definitions(-DOPENCL_BLOCKS_TEST=1)
 endif(OPENCL_FOUND)
 
-add_definitions(-DTEST_PATH="${CMAKE_SOURCE_DIR}/src/tests/")
+add_definitions(-DTEST_PATH="./tests/")
 
 
 add_executable(run_tests ${CMAKE_CURRENT_SOURCE_DIR}/test_main.cc)
